File: SeatingPlaceSuggestions/Java/SeatingPlaceSuggestions/src/main/java/org/weaveit/seatssuggestions/SeatingPlace.java
package org.weaveit.seatssuggestions;

import java.util.Objects;

public record SeatingPlace(String rowName, int number, PricingCategory pricingCategory, SeatingPlaceAvailability seatingPlaceAvailability) {

    public boolean isAvailable() {
        return seatingPlaceAvailability == SeatingPlaceAvailability.Available;
    }

    public boolean matchCategory(PricingCategory pricingCategory) {
        if (pricingCategory == PricingCategory.Mixed) {
            return true;
        }

        return this.pricingCategory == pricingCategory;
    }

    public SeatingPlace allocate() {
        if (seatingPlaceAvailability == SeatingPlaceAvailability.Available) {
            return new SeatingPlace(rowName, number, pricingCategory, SeatingPlaceAvailability.Allocated);
        }
        return this;
    }

    public boolean sameSeatingPlace(SeatingPlace seatingPlace) {
        return Objects.equals(rowName, seatingPlace.rowName) && number == seatingPlace.number;
    }

    @Override
    public String toString() {
        return rowName + number;
    }
}

File: SeatingPlaceSuggestions/CSharp/SeatingPlaceSuggestions/SeatingPlace.cs
using System.Collections.Generic;
using Value;

namespace SeatsSuggestions;

public class SeatingPlace(
    string rowName,
    int number,
    PricingCategory pricingCategory,
    SeatingPlaceAvailability seatingPlaceAvailability,
    int distanceFromTheMiddleOfTheRow = -1)
    : ValueType<SeatingPlace>
{
    public string RowName { get; } = rowName;
    public int Number { get; } = number;
    public PricingCategory PricingCategory { get; } = pricingCategory;
    public SeatingPlaceAvailability SeatingPlaceAvailability { get; } = seatingPlaceAvailability;
    public int DistanceFromTheMiddleOfTheRow { get; } = distanceFromTheMiddleOfTheRow;

    public bool IsAvailable()
    {
        return SeatingPlaceAvailability == SeatingPlaceAvailability.Available;
    }

    public override string ToString()
    {
        return $"{RowName}{Number}";
    }

    public SeatingPlace Allocate()
    {
        if (SeatingPlaceAvailability == SeatingPlaceAvailability.Available)
            return new SeatingPlace(RowName, Number, PricingCategory, SeatingPlaceAvailability.Allocated);

        return this;
    }

    public bool SameSeatingPlace(SeatingPlace seatingPlace)
    {
        return RowName == seatingPlace.RowName && Number == seatingPlace.Number;
    }

    protected override IEnumerable<object> GetAllAttributesToBeUsedForEquality()
    {
        return new object[] { RowName, Number, PricingCategory, SeatingPlaceAvailability };
    }

    public bool MatchCategory(PricingCategory pricingCategory)
    {
        if (PricingCategory.Mixed == pricingCategory)
        {
            return true;
        }
        return this.PricingCategory == pricingCategory;
    }
}